### PR TITLE
 OCPBUGS-6661, OCPBUGS-9464: Handle mTLS CRLs, and fix accidental CRL duplication

### PR DESCRIPTION
**Pull request by rfredette**
This PR makes the router parse the certificates supplied in ROUTER_MUTUAL_TLS_AUTH_CA, download CRLs from any CRL distribution points that are found, and write them to a file that is supplied to HAProxy for mTLS certificate validation.

It also includes a fix for an issue where, when a CRL is distributed in a certificate that is not its issuer, the CRL would sometimes be missing or two copies of the CRL would be present.

### CODE REVIEW

File: pkg/router/crl/crl.go [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+	var nextCRLUpdate time.Time
+	crls := make(map[string]*x509.RevocationList)
+	updated := false
+	now := time.Now()
+	for len(clientCAData) > 0 {
+		block, data := pem.Decode(clientCAData)
+		if block == nil {
+			break
+		}
+		clientCAData = data
```
Comment (frobware) [inline pkg/router/crl/crl.go:335]: In `clientCAData` are we only concerned with certificates?

```go
if block.Type != "CERTIFICATE" {
    return nil, time.Time{}, false, fmt.Errorf("client CA bundle contains a non-certificate PEM block")
}
```

File: pkg/router/template/router.go [hunk 2]
```diff
@@ -430,23 +439,29 @@ func (r *templateRouter) writeDefaultCert() error {
...
-		if err := r.watchVolumeMountDir(filepath.Dir(crlPath), reloadFn); err != nil {
-			log.V(0).Info("failed to establish watch on mTLS certificate directory", "error", err)
+		crl.ManageCRLs(caPath, caUpdateChannel, crlReloadFn)
+		caReloadFn := func() {
+			// Send signal to CRL management goroutine that client CA has been changed
+			caUpdateChannel <- struct{}{}
+		}
+		if err := r.watchVolumeMountDir(filepath.Dir(caPath), caReloadFn); err != nil {
+			log.V(0).Error(err, "failed to establish watch on mTLS certificate directory")
 			return nil
```
Comment (frobware) [inline pkg/router/template/router.go:465]: Why does this not return the error?

**Review by frobware — CHANGES_REQUESTED** [review]

File: pkg/router/template/router.go (hunk 2 above)
Comment (frobware) [reply pkg/router/template/router.go:465]: @Miciah I noticed that `writeDefaultCert` doesn't return an error for its watch on calls to `watchVolumeMountDir()` either - do you know of any history for why this returns nil?

File: images/router/haproxy/conf/haproxy-config.template [hunk 3]
```diff
@@ -290,8 +292,13 @@ frontend fe_sni
     {{- "" }} crt-list /var/lib/haproxy/conf/cert_config.map accept-proxy
     {{- with (env "ROUTER_MUTUAL_TLS_AUTH") }}
       {{- "" }} verify {{. }}
-    {{- with (env "ROUTER_MUTUAL_TLS_AUTH_CA") }} ca-file {{. }} {{ else }} ca-file /etc/ssl/certs/ca-bundle.trust.crt {{ end }}
-    {{- with (env "ROUTER_MUTUAL_TLS_AUTH_CRL") }} crl-file {{. }} {{ end }}
+      {{- if (ne (env "ROUTER_MUTUAL_TLS_AUTH_CRL") "") }}
+        {{- with (env "ROUTER_MUTUAL_TLS_AUTH_CA") }} ca-file {{ . }} {{- else }} ca-file /etc/ssl/certs/ca-bundle.trust.crt {{ end }}
```
[the remark this answers is not in the document]
Comment (rfredette) [reply images/router/haproxy/conf/haproxy-config.template:296]: It's not an intentional change; I will update this to get rid of that drift.

File: pkg/router/crl/crl.go (hunk 1 above)
Comment (rfredette) [reply pkg/router/crl/crl.go:335]: We should only care about certificates; any other PEM block type shouldn't matter. I think log & skip any non-certificate content is the way to go.

File: pkg/router/template/router.go (hunk 2 above)
Comment (frobware) [reply pkg/router/template/router.go:465]: If we were to return an error I'd propose we do that change in 4.14+ only for the moment as we want the backports to discretely fix OCPBUGS-6661 and OCPBUGS-9464 only.

File: images/router/haproxy/conf/haproxy-config.template (hunk 3 above)
[the remark this answers is not in the document]
Comment (rfredette) [reply images/router/haproxy/conf/haproxy-config.template]: Done

File: pkg/router/crl/crl.go (hunk 1 above)
Comment (rfredette) [reply pkg/router/crl/crl.go:337]: Done

File: pkg/router/crl/crl.go [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+
+// getHTTPCRL gets a certificate revocation list using the provided HTTP URL. Returns an error if the CRL could not be
+// downloaded, or if parsing the CRL fails.
+func getHTTPCRL(url string) (*x509.RevocationList, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("http.Get failed: %w", err)
+	}
+	defer resp.Body.Close()
+	crlBytes, err := io.ReadAll(resp.Body)
```
[the remark this answers is not in the document]
Comment (rfredette) [reply pkg/router/crl/crl.go:434]: Done

**Pull request merged**